inferno-core/src/Inferno/Infer.hs

```diff
@@ -1031,17 +1034,50 @@ infer expr =
                             LText _ -> typeText
                         )
                     POne _ p -> do
-                      (t, vars) <- mkPatConstraint p
+                      (t, vars, cs) <- mkPatConstraint p
                       meta <- lookupEnv patLoc $ Left oneHash
                       attachTypeToPosition patLoc meta {ty = (Set.empty, ImplType Map.empty $ t .-> TOptional t)}
-                      return (TOptional t, vars)
+                      return (TOptional t, vars, cs)
                     PEmpty _ -> do
                       meta <- lookupEnv patLoc $ Left emptyHash
                       let (_, ImplType _ t) = ty meta
                       attachTypeToPosition patLoc meta
-                      return (t, [])
+                      return (t, [], Set.empty)
+                    PArray _ [] _ -> do
+                      tv <- fresh
+                      let t = TArray tv
+                      attachTypeToPosition
+                        patLoc
+                        TypeMetadata
+                          { identExpr = patternToExpr $ bimap (const ()) (const ()) pattern,
+                            ty = (Set.empty, ImplType Map.empty $ t),
+                            docs = Nothing
+                          }
+                      return (t, [], Set.empty)
+                    PArray _ ((p, _) : ps) _ -> do
+                      (t, vars1, csP) <- mkPatConstraint p
+                      (vars2, csPs) <- aux t ps
+                      let inferredTy = TArray t
+                      attachTypeToPosition
+                        patLoc
+                        TypeMetadata
+                          { identExpr = patternToExpr $ bimap (const ()) (const ()) pattern,
+                            ty = (Set.empty, ImplType Map.empty $ inferredTy),
+                            docs = Nothing
+                          }
+                      return (inferredTy, vars1 ++ vars2, csP `Set.union` csPs)
```

Comment:
Type checking for array patterns is similar to that of tuple patterns, except every element of the array must have the same type (constraint `tIst'` below), so I modified `mkPatConstraint` to also return a set of type constraints.